mountpoint-s3-crt/src/auth/imds_client.rs

```diff
@@ -0,0 +1,126 @@
+//! A client for retrieving ec2 instance metadata.
+
+use crate::auth::auth_library_init;
+use crate::common::allocator::Allocator;
+use crate::common::error::Error;
+use crate::io::channel_bootstrap::ClientBootstrap;
+use crate::io::retry_strategy::RetryStrategy;
+use crate::CrtError;
+use mountpoint_s3_crt_sys::{
+    aws_byte_buf, aws_imds_client, aws_imds_client_get_instance_type, aws_imds_client_new, aws_imds_client_options,
+    aws_imds_client_release,
+};
+use std::ffi::{OsStr, OsString};
+use std::os::unix::prelude::OsStrExt;
+use std::ptr::NonNull;
+
+/// A client for instance metadata query.
+#[derive(Debug)]
+pub struct ImdsClient {
+    /// A pointer to the underlying `aws_imds_client`
+    inner: NonNull<aws_imds_client>,
+}
+
+///Configurations for creating a new [ImdsClient].
+#[derive(Debug, Default)]
+pub struct ImdsClientConfig {
+    /// The structure that can pass into the CRT's functions.
+    inner: aws_imds_client_options,
+
+    /// The [ClientBootstrap] to use to create connection to IMDS.
+    client_bootstrap: Option<ClientBootstrap>,
+
+    /// The [RetryStrategy] to use to reschedule failed requests to IMDS. This is reference counted,
+    /// so we only need to hold onto it until this [ImdsClientConfig] is consumed, at which point the
+    /// client will take ownership.
+    retry_strategy: Option<RetryStrategy>,
+}
+
+impl ImdsClient {
+    /// Create a new [ImdsClient]
+    pub fn new(allocator: &Allocator, config: ImdsClientConfig) -> Result<Self, Error> {
+        auth_library_init(allocator);
+
+        // SAFETY: `config` is moved into the [ImdsClient] on success, so `config.inner` (and the values
+        // inside) are guaranteed to live at least as long as this [ImdsClient] does. `allocator` is
+        // guaranteed to be a valid allocator because of the type-safe wrapper.
+        let inner = unsafe { aws_imds_client_new(allocator.inner.as_ptr(), &config.inner).ok_or_last_error()? };
```

Comment:
It was a bit hard to follow at first the safety comment, and maybe we are not calling out the real reason it is safe.

Does this capture it well? (I did not verify the explanation about the allocator.)

```suggestion
        // SAFETY: Struct memory referenced by `config.inner` will be managed by CRT, so it is safe to drop `ImdsClientConfig` passed into this method.
        // It is safe to pass in `allocator` as a pointer as it is only used during initialization of the client.
        let inner = unsafe { aws_imds_client_new(allocator.inner.as_ptr(), &config.inner).ok_or_last_error()? };
```

Reply:
The modified comments make sense to me. However, the key words `unsafe` here is to guarantee the reference to two objects must have the same lifetime as `ImdsClient`

1. >   // SAFETY: Struct memory referenced by `config.inner` will be managed by CRT, so it is safe to drop `ImdsClientConfig` passed into this method.

    * `config.inner` will be actually used by CRT, the object(in memory) is owned by client side but not CRT. When we instantiate  a `ImdsClient` object, the object itself owns the `config` along with its inner object, and when the object of `ImdsClient` get dropped, `config` will be dropped together.


2, > // It is safe to pass in `allocator` as a pointer as it is only used during initialization of the client.
    * It's true, the adds up for this statement is CRT will use `allocator` in variety of places like allocate memory for internal `ListNodes` which act as a node of the request queue or allocates memory for internal `imds_user_data` etc. Rust client owns the memory of  `allocator` itself.

 the code inside an unsafe block will access memory in a valid way because there is no memory leak, and the reference passed in is implemented and managed by in Rust, the reference is live as long as the `ImdsClient`.

Reply:
OK.

So for 1., we are saying that `aws_imds_client_new` will be using the config and then doesn't need it anymore, so its safe to drop all of it at the end of this `new` method. At the moment, the safety comment for number 1 isn't covering why its safe yet.

Reply:
This still isn't resolved for me. For 1., I don't understand why `config` is moved into `ImdsClient`. `new()` takes ownership, passes a pointer into unsafe CRT code. Why is it not dropped?

Reply:
Ohhhh, I know why you are confused about this comment, that was my bad. I keep thinking I have ImdsClient like this

```
/// A client for instance metadata query.
#[derive(Debug)]
pub struct ImdsClient {
    /// A pointer to the underlying `aws_imds_client`
    inner: NonNull<aws_imds_client>,
}
```
but actually defined as this

```
 /// A client for instance metadata query.                   
 #[derive(Debug)]                                            
 pub struct ImdsClient { 3 implementations                   
     /// A pointer to the underlying `aws_imds_client`       
     inner: NonNull<aws_imds_client>,                        
     config: ImdsClientConfig,                               
 }                                                           
```
and when we create `ImdsClient`, we pass in` ClientConfig` which return Self {inner, config} that is `config` moved in
```

Reply:
There shouldn't be any need to retain the `ImdsClientConfig` beyond the end of `ImdsClient::new`. The `aws_imds_client_new` bumps the reference count of everything it needs out of the config struct ([retry strategy](https://github.com/awslabs/aws-c-auth/blob/54f8d804120daab9e0d75a56a113a222b334d0f9/source/aws_imds_client.c#L179), [bootstrap](https://github.com/awslabs/aws-c-http/blob/0600662610aa871a11aebe6ed67a11997317cbef/source/connection_manager.c#L869)) and doesn't hold onto the pointer to the `aws_imds_client_options`. That's why this is safe.